### DISCUSSION TITLE
The Maths behind gesture: 'facing center'

### DISCUSSION THREAD

**Discussion by YaraAmin** [vladmandic/human#150] [Q&A] [answered]
Can you explain to me the logic and the maths behind facing centre gesture in this code :

`for (let i = 0; i < res.length; i++) {
    if (!res[i].annotations || !res[i].annotations.leftEyeIris || !res[i].annotations.rightEyeIris) continue;
    const sizeXLeft = res[i].annotations.leftEyeIris[3][0] - res[i].annotations.leftEyeIris[1][0];
    const sizeYLeft = res[i].annotations.leftEyeIris[4][1] - res[i].annotations.leftEyeIris[2][1];
    const areaLeft = Math.abs(sizeXLeft * sizeYLeft);

    const sizeXRight = res[i].annotations.rightEyeIris[3][0] - res[i].annotations.rightEyeIris[1][0];
    const sizeYRight = res[i].annotations.rightEyeIris[4][1] - res[i].annotations.rightEyeIris[2][1];
    const areaRight = Math.abs(sizeXRight * sizeYRight);

    let center = false;
    const difference = Math.abs(areaLeft - areaRight) / Math.max(areaLeft, areaRight);
    if (difference < 0.25) {
      center = true;
      gestures.push({ iris: i, gesture: 'facing center' });
    }`
    
Thank you

**Comment by vladmandic** [comment, marked as answer]
i'd consider actual value of `iris: facing center` gesture less important than similar one `face: facing center` which looks at the Z coordinates of corners of the eyes (not corners of the face as with higher angles face corner is likely not visible to start with)  

`iris: facing center` is weird, but i've tried different methods and this algorithm gave ok results on average  
it first calculates square area of each eye (`areaLeft` and `areaRight`) based on corners of the eye  
and then if difference in eye areas is less than 25% it states `facing center` - meaning both eyes are of similar size so thus face is not turned to either side  

you can think of `iris: facing center` as an intermediate result towards `iris: looking center`
method that states `iris: looking center` starts if `iris: facing center` is detected and then looks at difference between iris center and up/down/left/right corners of eyes and if iris is still in center, it states `iris: looking center`
